### PR TITLE
Fix condition of antijoin for ALL quantified subqueries

### PR DESCRIPTION
**Pull request by zellerh**
This is a backport of https://github.com/greenplum-db/gpdb/pull/11490.

It fixes the wrong result reported in
https://github.com/greenplum-db/gpdb/issues/10967

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by zellerh** [comment]
The original PR by @darthunix also has a fix in the GPDB code. I want to see whether this fails somewhere in ICW without it. @darthunix, hope you can review this PR and comment on the need of the GPDB fix - maybe you have a test case?

**Comment by darthunix** [comment]
@zellerh originally GPDB code modification was caused by ["unexpected expression type in NestLoopJoin qual"](https://github.com/greenplum-db/gpdb/blob/940cf6486a91245a4bf5584931e555f77a9c00eb/src/backend/executor/nodeNestloop.c#L659) error in my regression. The reason was in a new Distinct expression added by original commit to LASJ Not-In. As now all regression tests pass without any errors, I guess that the problem query uses some other plan alternative and we have to  re-invent such example from the scratch (LASJ Not-In + nested loop + distinct). I failed to make the one at the moment. May be there is some ORCA GUC/minidump trace flag to use an exact alternative we want? Then I believe it is possible to get a proper plan for the executor.

**Comment by zellerh** [comment]
Thanks, @darthunix, for looking into this. Hopefully I'm now able to produce a test case.

tl;dr: We are ok without a GPDB change for 6X.

In more detail, here is what I did:

- comment out the code for a correlated plan in `CXformSubqueryUnnest::Transform()`

```
DROP TABLE IF EXISTS t;
CREATE TABLE t(i INT) DISTRIBUTED RANDOMLY;

INSERT INTO t VALUES (1), (2), (3);

set optimizer_enable_hashjoin to off;
select disable_xform('CXformSimplifySelectWithSubquery');

EXPLAIN
SELECT *
FROM t
WHERE i NOT IN (SELECT t2.i FROM t t1 LEFT JOIN t t2 ON FALSE);

SELECT *
FROM t
WHERE i NOT IN (SELECT t2.i FROM t t1 LEFT JOIN t t2 ON FALSE);
```

This produces a nested LASJ with the IS DISTINCT FROM as the
join predicate. It executes and produces the correct result (no rows):

```
Physical plan:
+--CPhysicalMotionGather(master)   rows:3   width:34  rebinds:1   cost:904755252.990016   origin: [Grp:7, GrpExpr:4]
   +--CPhysicalLeftAntiSemiNLJoinNotIn   rows:3   width:34  rebinds:1   cost:904755252.989971   origin: [Grp:7, GrpExpr:3]
      |--CPhysicalTableScan "t" ("t")   rows:3   width:34  rebinds:1   cost:431.000019   origin: [Grp:0, GrpExpr:1]
      |--CPhysicalSpool (Streaming)   rows:4   width:64  rebinds:1   cost:882688.051252   origin: [Grp:4, GrpExpr:3]
      |  +--CPhysicalMotionBroadcast    rows:4   width:64  rebinds:1   cost:882688.051236   origin: [Grp:4, GrpExpr:2]
      |     +--CPhysicalLeftOuterNLJoin   rows:4   width:64  rebinds:1   cost:882688.050950   origin: [Grp:4, GrpExpr:1]
      |        |--CPhysicalTableScan "t" ("t")   rows:3   width:30  rebinds:1   cost:431.000019   origin: [Grp:1, GrpExpr:1]
      |        |--CPhysicalConstTableGet Columns: ["i" (16), "ctid" (17), "xmin" (18), "cmin" (19), "xmax" (20), "cmax" (21), "tableoid" (22), "gp_segment_id" (23)] Values: []   rows:0   width:34  rebinds:1   cost:0.000000   origin: [Grp:2, GrpExpr:1]
      |        +--CScalarConst (1)   origin: [Grp:3, GrpExpr:0]
      +--CScalarIsDistinctFrom (=)   origin: [Grp:11, GrpExpr:0]
         |--CScalarCmp (=)   origin: [Grp:9, GrpExpr:0]
         |  |--CScalarIdent "i" (0)   origin: [Grp:5, GrpExpr:0]
         |  +--CScalarIdent "i" (16)   origin: [Grp:8, GrpExpr:0]
         +--CScalarConst (0)   origin: [Grp:10, GrpExpr:0]

 i
---
(0 rows)
```

What I'm seeing here in the `splitJoinQualExpr` function is that
`exprstate->xprstate.type` is set to `T_FuncExprState`, with the actual
expression being a `T_DistinctExpr`. The query runs fine on 6X and my
assumption is that no GPDB change is needed for 6X.

Also, I found that we actually don't split the IS DISTINCT FROM expression,
since the `T_DistinctExpr` is neither an `OpExpr` nor a `FuncExpr`. This is
another piece of evidence that the code in https://github.com/greenplum-db/gpdb/pull/11490 is ok.

**Comment by darthunix** [comment]
@zellerh sorry for long response - we had notional holidays in Russia.

If I understand the previous comment right, you successfully enter [splitJoinQualExpr()](https://github.com/zellerh/gpdb/blob/940cf6486a91245a4bf5584931e555f77a9c00eb/src/backend/executor/nodeNestloop.c#L632) function with `exprstate->xprstate.type ==  T_FuncExprState` and actual expression `T_DistinctExpr` during execution. My tries with [all_subquery_null](https://github.com/zellerh/gpdb/tree/all_subquery_null) branch never lead to this function and I [have](https://github.com/zellerh/gpdb/blob/940cf6486a91245a4bf5584931e555f77a9c00eb/src/backend/executor/nodeNestloop.c#L525) `JOIN_LEFT` type with `exprstate->xprstate.type ==  T_ExprState` and actual expression `T_Const`. As a result the code never reaches [splitJoinQualExpr()](https://github.com/zellerh/gpdb/blob/940cf6486a91245a4bf5584931e555f77a9c00eb/src/backend/executor/nodeNestloop.c#L527) to face a problem with `exprstate->xprstate.expr->type == T_DistinctExpr` and ["unexpected expression type in NestLoopJoin qual"](https://github.com/zellerh/gpdb/blob/940cf6486a91245a4bf5584931e555f77a9c00eb/src/backend/executor/nodeNestloop.c#L659) error. Can you verify the example above, please - may be you've missed some GUCs? I want to get the same results as you do.

**Comment by hardikar** [comment]
@darthunix I just took a closer look at this PR and what went into master, and although it is not strictly incorrect, I feel there is a simpler solution to fix the issue.

Specifically, I noticed that just projecting a column from `t1` in the subquery produces the correct output.
```
SELECT * FROM t WHERE i NOT IN (
    SELECT t1.i FROM t AS t1 LEFT JOIN t AS t2 ON FALSE
);
 i 
---
(0 rows)
```

Comparing the plans produced for these two queries by ORCA, one difference I noticed was that when projecting the `t2.i` column the inner LOJ gets converted to an Inner Join. An Inner Join is NULL rejecting, so the Anti-Semi-Not-In-Join doesn't see any tuples from it inner child, producing the wrong results. But when the `t1.i` is projected, a LOJ is retained; so the Anti-Semi-Not-In-Join sees the NULLs from its inner child, and thus produces the right results (no tuples).
```
shardikar=# explain SELECT * FROM t WHERE i NOT IN (
    SELECT t1.i FROM t AS t1 LEFT JOIN t AS t2 ON FALSE
);
                                              QUERY PLAN                                              
------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..883119.05 rows=2 width=4)
   ->  Hash Left Anti Semi (Not-In) Join  (cost=0.00..883119.05 rows=1 width=4)
         Hash Cond: (t.i = t_1.i)
         ->  Seq Scan on t  (cost=0.00..431.00 rows=1 width=4)
         ->  Hash  (cost=882688.05..882688.05 rows=4 width=4)
               ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..882688.05 rows=4 width=4)
                     ->  Nested Loop Left Join  (cost=0.00..882688.05 rows=2 width=4)
                           Join Filter: true
                           ->  Seq Scan on t t_1  (cost=0.00..431.00 rows=1 width=4)
                           ->  Result  (cost=0.00..0.00 rows=0 width=1)
                                 One-Time Filter: false
 Optimizer: Pivotal Optimizer (GPORCA)
(12 rows)

Time: 45.174 ms
shardikar=# explain SELECT * FROM t WHERE i NOT IN (
    SELECT t2.i FROM t AS t1 LEFT JOIN t AS t2 ON FALSE
);
                                              QUERY PLAN                                              
------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..883119.03 rows=1 width=4)
   ->  Hash Left Anti Semi (Not-In) Join  (cost=0.00..883119.03 rows=1 width=4)
         Hash Cond: (t.i = "inner".i)
         ->  Seq Scan on t  (cost=0.00..431.00 rows=1 width=4)
         ->  Hash  (cost=882688.03..882688.03 rows=1 width=4)
               ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..882688.03 rows=1 width=4)
                     ->  Nested Loop  (cost=0.00..882688.03 rows=1 width=4)
                           Join Filter: true
                           ->  Seq Scan on t t_1  (cost=0.00..431.00 rows=1 width=1)
                           ->  Result  (cost=0.00..0.00 rows=0 width=4)
                                 One-Time Filter: false
 Optimizer: Pivotal Optimizer (GPORCA)
(12 rows)
```
Digging into the `CXformSelect2Apply`, I think this happens because `CSubqueryHandler::FRemoveAllSubquery()` produces a NULL rejecting filter on the LOJ (as you described in https://github.com/greenplum-db/gpdb/pull/11490) and this gets normalized into an inner join.

So the simple fix would be to *not* produce this filter, and instead place the predicate in the LAS-Apply-NotIn directly, since it knows how to handle the NULL semantics of NOT-IN queries already.

**Comment by hardikar** [comment]
I opened another PR with this implementation: https://github.com/greenplum-db/gpdb/pull/12067

**Comment by hardikar** [comment]
Closing this PR in favor of https://github.com/greenplum-db/gpdb/pull/12067
